meta-client/src/test/java/org/apache/helix/metaclient/impl/zk/TestZkMetaClientAsyncOperations.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    public void setReturnCodeWhenFinished(int idx, int returnCode) {
+      _returnCode[idx] = returnCode;
+    }
+
+    public void setStatWhenFinished(int idx, MetaClientInterface.Stat stat) {
+      _stats[idx] = stat;
+    }
+
+    public void setDataWhenFinished(int idx, String data) {
+      _data[idx] = data;
+    }
+  }
+
+  @Test
+  public void testAsyncCreateSetAndGet() {
+    asyncContext[0] = new TestAsyncContext(2);
+    try (ZkMetaClient<String> zkMetaClient = createZkMetaClient()) {
+      zkMetaClient.connect();
+
+      zkMetaClient
+          .asyncCreate(entryKey, "async_create-data", MetaClientInterface.EntryMode.PERSISTENT,
+              new AsyncCallback.VoidCallback() {
+                @Override
+                public void processResult(int returnCode, String key) {
+                  asyncContext[0].setReturnCodeWhenFinished(0, returnCode);
+                  asyncContext[0].countDown();
+                }
+              });
+
+      zkMetaClient.asyncCreate(nonExistsEntry, "async_create-data-invalid",
+          MetaClientInterface.EntryMode.PERSISTENT, new AsyncCallback.VoidCallback() {
+            @Override
+            public void processResult(int returnCode, String key) {
+              asyncContext[0].setReturnCodeWhenFinished(1, returnCode);
+              asyncContext[0].countDown();
+            }
+          });
+
+      asyncContext[0].getCountDownLatch().await();
+
+      Assert.assertEquals(asyncContext[0].getReturnCode(0), KeeperException.Code.OK.intValue());
+      Assert.assertEquals(asyncContext[0].getReturnCode(1), KeeperException.Code.NONODE.intValue());
+
+      // create the entry again and expect a duplicated error code
+      asyncContext[0] = new TestAsyncContext(1);
+      zkMetaClient
+          .asyncCreate(entryKey, "async_create-data", MetaClientInterface.EntryMode.PERSISTENT,
+              new AsyncCallback.VoidCallback() {
+                @Override
+                public void processResult(int returnCode, String key) {
+                  asyncContext[0].setReturnCodeWhenFinished(0, returnCode);
+                  asyncContext[0].countDown();
+                }
+              });
+      asyncContext[0].getCountDownLatch().await();
+      Assert.assertEquals(asyncContext[0].getReturnCode(0),
+          KeeperException.Code.NODEEXISTS.intValue());
+
+
+      // test set
+      asyncContext[0] = new TestAsyncContext(1);
+      zkMetaClient
+          .asyncSet(entryKey, "async_create-data-new", 0,
+              new AsyncCallback.StatCallback() {
+                @Override
+                public void processResult(int returnCode, String key,
+                    @Nullable MetaClientInterface.Stat stat) {
+                  asyncContext[0].setReturnCodeWhenFinished(0, returnCode);
+                  asyncContext[0].setStatWhenFinished(0, stat);
+                  asyncContext[0].countDown();
+                }
+              });
+      asyncContext[0].getCountDownLatch().await();
```

Comment:
nit: better to set a timeout and Assert on that

Reply:
TFTR. Updated.